packages/next-intl/src/middleware/middleware.tsx

```diff
@@ -24,14 +24,15 @@ function receiveConfig(config: MiddlewareConfig) {
   return result;
 }
 
+type MiddlewareResponseInit = Parameters<typeof NextResponse.next>[0]
 export default function createMiddleware(config: MiddlewareConfig) {
   const configWithDefaults = receiveConfig(config);
 
   // Currently only in use to enable a seamless upgrade path from the
   // `{createIntlMiddleware} from 'next-intl/server'` API.
   const matcher: Array<string> | undefined = (config as any)._matcher;
 
-  return function middleware(request: NextRequest) {
+  return function middleware(request: NextRequest, _?: NextFetchEvent, responseInit?: MiddlewareResponseInit) {
```

Comment:
Hmm, this doesn't look quite right. You're accepting a third parameter now for the middleware, but the goal is to keep existing incoming headers, right?

I think these tests could help to verify the implementation:

```tsx
it('retains request headers for the default locale', () => {
  middleware(
    createMockRequest('/', 'en', 'http://localhost:3000', undefined, {
      'x-test': 'test'
    })
  );
  expect(
    MockedNextResponse.rewrite.mock.calls[0][1].request.headers.get(
      'x-test'
    )
  ).toBe('test');
});

it('retains request headers for secondary locales', () => {
  middleware(
    createMockRequest('/de', 'de', 'http://localhost:3000', undefined, {
      'x-test': 'test'
    })
  );
  expect(
    MockedNextResponse.next.mock.calls[0][0].request.headers.get('x-test')
  ).toBe('test');
});
```

These tests are set up to be placed at the spot in the test suite where you've added your test!

I've added another parameter for `createMockRequest` here to accept `extraHeaders?: Record<string, string>`.

If the goal of the consumer is to modify request headers, this can be done [before the middleware is invoked](https://next-intl-docs.vercel.app/docs/next-13/middleware#composing-other-middlewares).

Does this help? Let me know if I can help!

Reply:
@amannn My initial intention was to provide ability to set custom request headers that can be consumed within server components. Previous implementation did it by implicitly forwarding headers from incoming request. But my concern is that it could go wrong at some point and such explicit setting of the `responseInit` looks more robust and extendable. That is how I see adjusted `withExtraMiddleware` example:
``` typescript
import createMiddleware, { NextMiddleware } from 'next-intl/middleware';
import { NextMiddlewareResult } from 'next/dist/server/web/types';
import {NextFetchEvent, NextRequest, NextResponse} from 'next/server';

export default withExtraMiddleware(
  createMiddleware({
    locales: ['en', 'de'],
    defaultLocale: 'en'
  })
);

type MiddlewareResponseInit = Parameters<typeof NextResponse.next>[0]

function withExtraMiddleware(next: NextMiddleware) {
  return async (request: NextRequest, event: NextFetchEvent) => {
    // Step 1: Potentially change the incoming request
    request.headers.set('x-test', 'test');

    // Step 2: Call the nested next-intl middleware
    const response = next(request, event, { request: { headers: request.headers } });

    // Step 3: Potentially change the response
    if (response) {
      response.headers.set('x-test', 'test');
    }

    return response;
  };
}

export const config = {
  // Skip all paths that should not be internationalized
  matcher: ['/((?!_next|.*\\..*).*)']
};
```

Thank you for the tests!

Reply:
Thanks for explaining @ARochniak!

> Previous implementation did it by implicitly forwarding headers from incoming request. But my concern is that it could go wrong at some point.

Do you have anything particular in mind that could go wrong? I think that by adding tests we can guarantee that this continues to work in the future.

Generally, if something can be solved with the existing APIs, I wouldn't want to add features. By doing this, we have to maintain less features within the library and also for the consumer there is a single way to achieve a particular goal (if we'd support a third argument, it's questionable what happens to existing request headers for example).




Therefore I'd really be in favor of not adding more API than necessary. Hope this makes sense!

Reply:
@amannn Thanks for clarification. My concern is mostly based on some "God feeling" 😊. I haven't any issue about using previous approach, and I just wanted to discuss with you different option.